modules/webgl/src/adapter/resources/webgl-command-buffer.ts

```diff
@@ -171,9 +173,12 @@ function _copyTextureToBuffer(device: WebGLDevice, options: CopyTextureToBufferO
     const sourceWidth = width || framebuffer.width;
     const sourceHeight = height || framebuffer.height;
 
-    // TODO - hack - should be deduced
-    const sourceFormat = GL.RGBA;
-    const sourceType = GL.UNSIGNED_BYTE;
+    const format = TEXTURE_FORMATS[framebuffer.texture.format];
+    assert(format.dataFormat, `Unsupported format, "${framebuffer.texture.format}"`);
+    assert(format.types, `Unsupported format, "${framebuffer.texture.format}"`);
+
+    const sourceFormat = format.dataFormat;
```

Comment:
Just out of curiousity, I have been wondering how much liberty the user has to specify a different source format in this API.  e.g. `GL.R` source data for a RGBA texture (that would presumably only set the first component of each color). 

Since the API allows this to be specified separately from the texture format it seems that there would be room for some variation but maybe it is just for some very limited cases.

Reply:
Not sure about WebGPU — but in WebGL we get this error to select `GL.RED` from an RGBA texture:

```
GL_INVALID_OPERATION: Invalid format and type combination.
```

Reply:
I don't think this level of flexibility exists in WebGPU.

Thanks for testing, in WebGL I suspect it is probably just there to handle a few special cases. 

There is probably a table in the spec somewhere, but let's not worry about that.